xwin.c

```diff
@@ -1895,6 +1959,146 @@ set_wm_client_machine(Display * dpy, Window win)
 	XSetWMClientMachine(dpy, win, &tp);
 }
 
+static long parse_single_word_mods(const char *str)
+{
+	int i;
+	size_t len;
+	size_t pat_len;
+
+	if (!str) return -1;
+
+	len = strlen(str);
+
+	if (!len) return -1;
+
+	for (i = 0; i < NUM_MODS; i++) {
+
+		pat_len = strlen(parse_mods[i].both_str);
+
+		if (pat_len == len) {
+			if (!strncmp(str, parse_mods[i].both_str, pat_len)) {
+				return parse_mods[i].both_mod;
+			}
+		}
+
+		pat_len = strlen(parse_mods[i].any_str);
+
+		if (pat_len == len) {
+			if (!strncmp(str, parse_mods[i].any_str, pat_len)) {
+				return parse_mods[i].any_mod;
+			}
+		}
+
+		pat_len = strlen(parse_mods[i].right_str);
+
+		if (pat_len == len) {
+			if (!strncmp(str, parse_mods[i].right_str, pat_len)) {
+				return parse_mods[i].right_mod;
+			}
+		}
+
+		pat_len = strlen(parse_mods[i].left_str);
+
+		if (pat_len == len) {
+			if (!strncmp(str, parse_mods[i].left_str, pat_len)) {
+				return parse_mods[i].left_mod;
+			}
+		}
+
+	}
+
+	return -1;
+}
+
+static long parse_mod_request(const char *str)
+{
+	int i;
+	long bv;
+	long tmp_bv;
+	long sec_bv;
+
+	long idx;
+
+	char *p;
+
+	char *dup_str;
+
+	p = strchr(str, '_');
+
+	if (!p) {
+		tmp_bv = parse_single_word_mods(str);
+
+		if (tmp_bv == -1) {
+			/* Stick to defaults */
+			return -1;
+		} else {
+			return tmp_bv;
+		}
+	} else {
+		idx = (long)p - (long)str;
+
+		dup_str = strdup(str);
+
+		if (!dup_str) {
+			perror("strdup");
+			return -1;
+		}
+
+		dup_str[idx] = 0;
+
+		tmp_bv = parse_single_word_mods(dup_str);
+		sec_bv = parse_single_word_mods(&dup_str[idx + 1]);
+
+		if ((tmp_bv == -1) || (sec_bv == -1)) return -1;
+
+		if (tmp_bv == sec_bv) return -1;
+
+		for (i = 0; i < NUM_MODS; i++) {
+
+			if (tmp_bv == (mods[i].left_mod | mods[i].right_mod)) {
+				return -1;
+			}
+
+			if (sec_bv == (mods[i].left_mod | mods[i].right_mod)) {
+				return -1;
+			}
+		}
+
+		for (i = 0; i < NUM_MODS; i++) {
+			if (tmp_bv == mods[i].any_mod) {
+				if ((sec_bv == mods[i].right_mod) || (sec_bv == mods[i].left_mod)) {
+					return -1;
+				}
+			}
+
+			if (sec_bv == mods[i].any_mod) {
+				if ((tmp_bv == mods[i].right_mod) || (tmp_bv == mods[i].left_mod)) {
+					return -1;
+				}
+			}
+
+		}
+
+		return (tmp_bv | sec_bv);
+	}
+}
+
+int xwin_parse_grab_combo(const char *str)
+{
+	long bv;
+
+	bv = parse_mod_request(str);
+
+	if (bv != -1) {
+		g_grab_ctx.mod_supplied = bv;
+		g_grab_ctx.mod_allowed = g_grab_ctx.mod_supplied;
+		return 0;
+	} else {
+		logger(Keyboard, Warning, "%s(): Failed to parse supplied ungrab key combo (%s). Setting to default (any_ctrl + any_alt)", __func__, str);
```

Comment:
Would it be possible to have this say `ctrl_alt` instead of `any_ctrl + any_alt` to prevent confusion about the key combo format?